finance_dl/gemini.py

```diff
@@ -0,0 +1,289 @@
+
+- `credentials`: Required.  Must be a `dict` with `'api_key'`, `'api_secret'` keys.
+
+- `output_directory`: Required.  Must be a `str` that specifies the path on the
+  local filesystem where the output will be written.  If the directory does not
+  exist, it will be created.
+
+Output format:
+==============
+
+CSV files.
+
+output_directory
+-- trades.1632852595934.csv
+-- transfers.1632752595924.csv
+-- balances.2021-06-01.csv
+-- balances.2021-07-01.csv
+...
+
+The timestamp in the trades/transfers files indicates the timestamp of the last trade/transfer contained in the file and is used to download only later ones.
+
+Example:
+========
+
+    def CONFIG_gemini():
+        return dict(
+            module='finance_dl.gemini',
+            credentials={
+                'key': 'XXXXXX',
+                'secret': 'XXXXXX',
+            },
+            output_directory=os.path.join(data_dir, 'gemini'),
+        )
+"""
+
+import io
+import urllib.parse
+import re
+import dateutil.parser
+import datetime
+import logging
+import os, time, shutil, glob, re
+
+import requests
+import json
+import base64
+import hmac
+import hashlib
+import datetime, time
+import csv
+
+logger = logging.getLogger('gemini_downloader')
+
+BALANCES_URL = "https://api.gemini.com/v1/balances"
+TRADES_URL = "https://api.gemini.com/v1/mytrades"
+TRANSFERS_URL = "https://api.gemini.com/v1/transfers"
+TICKERS_URL = "https://api.gemini.com/v2/ticker"
+
+
+def get_request_headers(api_key, b64, signature):
+    return   {
+        'Content-Type': "text/plain",
+        'Content-Length': "0",
+        'X-GEMINI-APIKEY': api_key,
+        'X-GEMINI-PAYLOAD': b64,
+        'X-GEMINI-SIGNATURE': signature,
+        'Cache-Control': "no-cache"
+        }
+
+
+class RateLimitedRetryingRequester:
+    """
+    For public API entry points, we limit requests to 120 requests per minute, and recommend that you do not exceed 1 request per second.
+    When requests are received at a rate exceeding X requests per minute, we offer a "burst" rate of five additional requests that are queued but their processing is delayed until the request rate falls below the defined rate.
+
+When you exceed the rate limit for a group of endpoints, you will receive a 429 Too Many Requests HTTP status response until your request rate drops back under the required limit.
+    """
+    MIN_PERIOD = 0.5
+    last_req_ts = 0
+    RETRY_CODES = [429]
+    NUM_RETRIES = 3
+
+    def make_request(self, url, headers = None, get = True):
+        now = time.time()
+        wait = self.MIN_PERIOD - (now - self.last_req_ts)
+        if wait > 0:
+            time.sleep(wait)
+        try:
+            r = requests.get(url) if get else requests.post(url, headers=headers)
+            self.last_req_ts = time.time()
+            r.raise_for_status()
+            return r.json()
+        except requests.HTTPError as exc:
+            if exc.response.status_code not in self.RETRY_CODES:
+                raise exc
```

Comment:
I think this method could be refactored to avoid duplicating this block of code (if nothing else, by extracting it to a separate method), but I don't feel strongly about it.